Wrapper to MonIT Grafana APIs

amaltaro opened the pull request:
Fixes #9532 
Fixes #9572

#### Status
In development

#### Description
Changes are:
* a new mandatory secrets variable is required now `GRAFANA_TOKEN`.
* the `DASHBOARD_URL` secrets var has been deprecated
* created a new service class to interact with MonIT Grafana (it's does not inherit from the `Services` super class because it relies on tokens instead of x509 certs)
* created an specific client API to retrieve the old/deprecated SSB metrics, such as:
  * site status
  * CPU bound slots
  * IO bound slots
* updated the agent configuration with new parameters for the grafana url, api and token (further changes to the deployment script are required)
* last but not least, updated AgentStatusWatcher thread to use Grafana instead of SSB

#### Is it backward compatible (if not, which system it affects?)
no, it's a new feature

#### Related PRs
none

#### External dependencies / deployment changes
Deployment changes: https://github.com/dmwm/deployment/pull/873

amaltaro committed:
Wrapper to MonIT Grafana APIs

fix inheritance and how to pass headers

Remake class with direct reference to pycurl module

reorganize a few things

remove my debugging

rename module

vkuznet commented:
Alan, let me ask WHY DO YOU DEVELOP WMCore class for monitoring rather than using CMSMonitoring tools? We already provided [monit.py](https://github.com/dmwm/CMSMonitoring/blob/master/src/python/CMSMonitoring/monit.py) for querying Monit ES and InfluxDB instances. Why do you need your own implementation? Since this is monitoring task I rather suggest that you drop your own codebase and we move everything you need to CMSMonitoring where it should belong. Please outline clearly your requirements and let's work on code where it should belong.

amaltaro committed:
Replace SSB by Grafana in AgentStatusWatcher

hardwire grafana URL in the configuration file; grafana placehold in the secrets template

remove dashboard_url; add grafana_token arg to wmagent mod bin

break token in two parts

add secrets var to the testing setup

amaltaro commented:
apparently I got everything in place. @todor-ivanov please review, such that we can backport and patch all the production agents.

jenimal commented:
Does this mean we are close to having WmCore over to the new monitoring 
so we aren't using the wrong stats?
You Rock Alan!  And welcome back!

Jen

On 3/13/20 9:38 AM, Alan Malta Rodrigues wrote:
>
> apparently I got everything in place. @todor-ivanov 
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_todor-2Divanov&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=FHpiIPqkXGBwuHA_iRHdSQ&m=HhCKn44yR-Y11u3grERud4kFedWK6JUNRgszsZV9Rvo&s=lnQVUczgYWwq1pXpROotEuIgFOMVTPVPHmf-uf1gUnk&e=> 
> please review, such that we can backport and patch all the production 
> agents.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_pull_9555-23issuecomment-2D598751650&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=FHpiIPqkXGBwuHA_iRHdSQ&m=HhCKn44yR-Y11u3grERud4kFedWK6JUNRgszsZV9Rvo&s=VTOUQSW9yC-CGN2SH3tiQs68srnNqYsb98tL93O4lBg&e=>, 
> or unsubscribe 
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABFSFW7BQYTHXHL4SFXIPOLRHJANZANCNFSM4LG5MP3Q&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=FHpiIPqkXGBwuHA_iRHdSQ&m=HhCKn44yR-Y11u3grERud4kFedWK6JUNRgszsZV9Rvo&s=Pfi5OxocLiNRun5ravYHbL1ADhuE7vtmrFfesjiFoHw&e=>.
>

amaltaro commented:
Yes, Jen. I think this one is ready to be backported and applied to all the agents. I'm just waiting for Todor's feedback here.

BTW, here is the log of changes when I patch one of the testbed agents:
```
2020-03-13 17:41:09,113:139891446904576:INFO:ResourceControlUpdater:Changing T2_BE_UCL state from Draining to Normal
2020-03-13 17:41:09,400:139891446904576:INFO:ResourceControlUpdater:Changing T2_RU_INR state from Draining to Normal
2020-03-13 17:41:09,407:139891446904576:INFO:ResourceControlUpdater:Changing T2_IT_Pisa state from Draining to Normal
2020-03-13 17:41:09,412:139891446904576:INFO:ResourceControlUpdater:Changing T2_US_Florida state from Normal to Draining
2020-03-13 17:41:09,417:139891446904576:INFO:ResourceControlUpdater:Changing T2_ES_CIEMAT state from Draining to Normal
2020-03-13 17:41:09,422:139891446904576:INFO:ResourceControlUpdater:Changing T2_RU_ITEP state from Normal to Draining
2020-03-13 17:41:09,853:139891446904576:INFO:ResourceControlUpdater:Updating T2_TW_NCHC site thresholds for pend/runn: 822/1096
2020-03-13 17:41:10,036:139891446904576:INFO:ResourceControlUpdater:Updating T1_FR_CCIN2P3 site thresholds for pend/runn: 5373/7165
2020-03-13 17:41:10,089:139891446904576:INFO:ResourceControlUpdater:Updating T1_DE_KIT site thresholds for pend/runn: 5811/7748
2020-03-13 17:41:10,182:139891446904576:INFO:ResourceControlUpdater:Resource control cycle finished updating site state and thresholds.
2020-03-13 17:41:10,183:139891446904576:INFO:BaseWorkerThread:ResourceControlUpdater took 1.270 secs to execute
```

amaltaro commented:
@vkuznet Valentin, if it wasn't for Todor, I'd have missed your comment here.
As you probably noticed, I develop this module because I wasn't aware there was such functionality already in place. It would have been better if you had made such comment in the GH issue, such that we could consider it since the beginning.

It's never too late to do the right thing, so feel free to put a GH issue in place and we will revisit it in the near future. Right now, this patch is ready and production agents need it for "yesterday".
Thanks for your comment though.

vkuznet commented:
Alan, I posted on this very issue my comment, I don't know why you no longer read the comments. I cannot follow all newly open issues, and especially I cannot foresee when and how you will require feedback about monitoring but at least you are aware that such group exists and its responsibilities.

Anyway, followed your suggestion I opened the separate github issue to outline the issue.

todor-ivanov commented on src/python/WMCore/Services/MonIT/Grafana.py:
```diff
@@ -0,0 +1,133 @@
+        self.logger = self.configDict.get("logger", logging.getLogger())
+        self.logger.info("MonIT service initialized with parameters: %s", self.configDict)
+
+    def updateToken(self, token):
```
Where is that one called? What should be the frequency of token refreshments? Is it supposed to be an external service which will take care of that?

amaltaro replied:
It's not called anywhere, yet. However, it might be needed in the future (if we keep using this module)

todor-ivanov reviewed (commented):
A few comments are put in the code. But those were nothing major. The code looks good even without them being worked on.

amaltaro commented on src/python/WMCore/Services/MonIT/Grafana.py:
```diff
@@ -0,0 +1,133 @@
+        if not queryStr:
+            ### NOTE: these '\n' new lines are mandatory to get it working...
+            queryStr = '{"search_type":"query_then_fetch","ignore_unavailable":true,"index":["%s"]}\n' % indexName
+            queryStr += '{"size":500,"query":{"bool":{"filter":[{"range":{"metadata.timestamp":{"gte":"now-1d","lte":"now","format":"epoch_millis"}}},' \
```
I agree :)
I had a hard time to get the correct format/syntax working for this query.
I'll try to improve this query on Monday (only on the master branch though).

amaltaro commented on src/python/WMCore/Services/MonIT/Grafana.py:
```diff
@@ -0,0 +1,133 @@
+        # TODO: allow flexible time range instead of the hardwired last 24h
+        # TODO: allow flexible number of rows, instead of the hardwired 500
+        results = {}
+        if metricName not in ('prod_status', 'core_cpu_intensive', 'core_io_intensive',
```
Agreed! I'm gonna make it an instance variable.

amaltaro commented on src/python/WMCore/Services/MonIT/Grafana.py:
```diff
@@ -0,0 +1,133 @@
+    # PY3
+    from urllib.parse import urljoin
+
+class Grafana(object):
```
That's more like your and my taste ;) If you look into the other wrappers in the Services package, you'll see they are usually named after the service we need to interact.

amaltaro commented:
Thanks for your review, Todor. I'll get some of those fixed on Monday. I also need to implement the deployment logic, which will take those new parameters from the secrets file and put them in the agent config.

amaltaro commented:
@todor-ivanov @khurtado could you please help me validating this patch? It requires deployment (and secrets) changes, as in the PR description.

I'm trying to redeploy the agent on vocms0262, but I'm getting a weird error that I can't figure out where it comes from:
```
*** Initializing the agent ***
Initialising Agent...
Error modifying configuration:
 'central_logdb_url'
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.3.0/sw/slc7_amd64_gcc630/cms/wmagent/1.3.0/bin/wmagent-mod-config", line 343, in main
    cfg = modifyConfiguration(cfg, **parameters)
  File "/data/srv/wmagent/v1.3.0/sw/slc7_amd64_gcc630/cms/wmagent/1.3.0/bin/wmagent-mod-config", line 114, in modifyConfiguration
    config.General.central_logdb_url = args["central_logdb_url"]
KeyError: 'central_logdb_url'
```

I have tried it many times, with minor changes, and still I have no clue why it complains about this argument...

amaltaro commented:
The issue seems to be related to a blank space in one of those variables. So that token is "Bearer blahblah", which makes that script to get lost and break the deployment... Debugging it further to check what can be done.

amaltaro commented:
I'm done with my changes and tests. Everything seems to be working fine now, including the deployment step (mind the new option)

amaltaro commented:
Merging it to get the updates in jenkins. If needed, I will follow up with a further PR.
@todor-ivanov it would still be interesting to have your review here.

todor-ivanov reviewed (commented):
@amaltaro , Sorry I forgot to approve your changes in the review, and just made a comment last time. This looks good to me now.
